### ISSUE TITLE
Add validation that `--emulation-version=<version>` >= `DefaultKubeBinaryVersion`-3 

### ISSUE THREAD

**Issue by aaron-prindle** [kubernetes/kubernetes#127514]
### What would you like to be added?

Currently we have validation logic here:

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/version/version.go#L171-L180

```
// ValidateKubeEffectiveVersion validates the EmulationVersion is equal to the binary version at 1.31 for kube components.
// emulationVersion is introduced in 1.31, so it is only allowed to be equal to the binary version at 1.31.
func ValidateKubeEffectiveVersion(effectiveVersion EffectiveVersion) error {
	binaryVersion := version.MajorMinor(effectiveVersion.BinaryVersion().Major(), effectiveVersion.BinaryVersion().Minor())
	if binaryVersion.EqualTo(version.MajorMinor(1, 31)) && !effectiveVersion.EmulationVersion().EqualTo(binaryVersion) {
		return fmt.Errorf("emulation version needs to be equal to binary version(%s) in compatibility-version alpha, got %s",
			binaryVersion.String(), effectiveVersion.EmulationVersion().String())
	}
	return nil
}
```

That makes sure that `--emulation-version=<version>` `version` that is passed is >=1.31.  Currently this check is enough as k8s latest release is 1.31 but the actual restriction for `--emulation-version` is the current k8s binary version-3 (denoted as `n-3`).  As such the validation rule needs to be updated to prevent unsupported emulation versions for k8s versions `v1.35+`

Related PR comment denoted this future work: https://github.com/kubernetes/kubernetes/pull/126977#discussion_r1746105686

### Why is this needed?

This is needed as w/o this, users could set --emulation-version=\<n-4\> in the future (eg:  k8s versions `v1.35+`) which is not supported and could break them

/sig api-machinery

**Comment by omerap12** [comment]
I would like to work that if that's ok :) 
/assign

**Comment by liggitt** [comment]
Agree with not letting it be greater.

I'm not sure we're ready to relax to -3 yet… what does that mean for features which have graduated in the last 3 releases, locked to true, and whose conditional logic has been removed already?

**Comment by Jefftree** [comment]
> Agree with not letting it be greater.
> 
> I'm not sure we're ready to relax to -3 yet… what does that mean for features which have graduated in the last 3 releases, locked to true, and whose conditional logic has been removed already? We can't actually emulate those versions, right?

We set the minimum emulation version to 1.31, so features that graduated <=1.31 are unaffected unless `LockToDefault` is set to false (We don't have any at the moment).

Going forward, yes we need to make sure that if features have graduated in the last 3 releases, they should not be removed until `n-3` is satisfied. This means features should be removed at least 3 releases after graduation. I see convention that already aims for 3 releases but we would probably need to update the [sig arch documentation](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/feature-gates.md#ga-features) to enforce it since the official policy is 2 releases at the moment? Otherwise yeah we'll have inaccuracies emulating those versions.

**Comment by liggitt** [comment]
> we need to make sure that if features have graduated in the last 3 releases, they should not be removed until `n-3` is satisfied. This means features should be removed at least 3 releases after graduation. I see convention that already aims for 3 releases

there's no expectation that the code the gate protects remains after GA for three releases... in fact, we have the opposite expectation... that all gates that reach GA / locked on immediately have all conditional non-enabled code branches removed.

**Comment by Jefftree** [comment]
> there's no expectation that the code the gate protects remains after GA for three releases... in fact, we have the opposite expectation... that all gates that reach GA / locked on immediately have all conditional non-enabled code branches removed.

Hm I see that's a bit unfortunate. So from my understanding: Every GA promotion generally has the `// remove in v1.x` comment where x is approximately the current release + 3. This references specifically the gates, and the enablement logic may not follow. 

Per the [documentation](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/feature-gates.md#ga-features)

> [After at least two releases post-GA and deprecation](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecation), feature gates should be removed. Typically, we add a comment in [the code](https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go) such as: // remove in 1.23 to signal when we plan to remove the feature gate. We provide this grace period to give users time to stop referencing "finished" gates. If a feature gate is removed and a user has forgotten to drop the reference to it (e.g. in the CLI flags of kube-apiserver), then they will see a hard failure.

> When we set LockToDefault to true, we also remove all references to the feature gate from the codebase.

when a feature is GA'd, the recommendation is to preserve the gate for a couple releases while removing the reference logic immediately. So this will definitely break our ability to handle emulating features at previous versions. We will need both the gate and enablement logic to be preserved for the full 3 releases to support accurate emulation.

On another note, we are trying to handle preserving non-enabled code branches in tests https://github.com/kubernetes/kubernetes/issues/127477 after a feature goes GA.

**Comment by liggitt** [comment]
> We will need both the gate and enablement logic to be preserved for the full 3 releases to support accurate emulation.

Yeah, that's not well understood today, isn't enforced by anything yet, and will end up leaving 99% dead code branches for all features for an extra year across all sigs :-/

I think that was all known in advance on this feature, but I don't think ~anyone outside the folks working on this feature understand that impact or are prepared for it.

**Comment by Jefftree** [comment]
Fixed by https://github.com/kubernetes/kubernetes/pull/128279.

/close
